File: Georechner_C++/daten/strecke.cpp
#include <iostream>
#include "mainwindow.h"
#include <typeinfo>

Strecke::Strecke(const Punkt& p_p1, const Punkt& p_p2, const int& p_d, const std::string& p_dnr): m_p1(p_p1), m_p2(p_p2), m_d(p_d), m_dnr(p_dnr) {}

Strecke::Strecke(const Punkt& p_p1, float& p_laenge, const std::string& p_dnr ): m_p1(p_p1), m_laenge(p_laenge), m_dnr(p_dnr){
    std::cout<<"Konstruktor 2 aufgerufen"<<std::endl;
    std::string l_str = std::to_string(this->m_laenge);
    int m_d = sizeof(l_str.substr(l_str.find(".") + 1))/16;

    this->m_p2 = Punkt(this->m_p1.hole_y(), this->m_p1.hole_x() + this->m_laenge);
    std::cout<<m_p1.str()<<"   "<<m_p2.str()<<std::endl;
}

void Strecke::setze_nr(std::string& set_dnr){
    this->m_dnr = set_dnr;}

std::string Strecke::hole_nr(){
    return this->m_dnr;}

Punkt Strecke::hole_p1(){
    return this->m_p1;}
Punkt Strecke::hole_p2(){
        return this->m_p2;}

std::array<float, 2> Strecke::riwi_laenge(){
        float t12 = atan2((this->m_p2.hole_y() - this->m_p1.hole_y()), (this->m_p2.hole_x() - this->m_p1.hole_x()));
        float s12 = sqrt(pow(this->m_p2.hole_y() - this->m_p1.hole_y(), 2) + pow(this->m_p2.hole_x() - this->m_p1.hole_x(), 2));
        std::cout<<m_p1.str()<<" in riwi laenge  "<<m_p2.str()<<std::endl;        //s12 = roundoff(s12, 5);
        t12 = rad2gon(t12);
        std::array<float, 2> arr_t12_s12 = {t12, s12};
        return arr_t12_s12;
}

std::string Strecke::str_d(){
    std::array<float, 2> ts = this->riwi_laenge();
    std::string zeichen = "Nr:" + this->m_dnr + "p1:" + this->m_p1.str() + ", p2:" + this->m_p2.str() + ", t=" + std::to_string(ts[0]) + ", s=" + std::to_string(ts[1]);
    return zeichen;
}

File: Python/daten/strecke.py
import copy
import grundlagen.zweitegrund
from typing import Tuple
import daten.punkt as pkt

class Strecke:
    """Klasse Strecke
    """
    def __init__(self,p_p1: pkt.Punkt,p_p2: pkt.Punkt,p_d=20, p_nr=""):
        """
        :param p_p1: Anfangspunkt
        :type: daten.punkt.Punkt
        :param p_p2: Endpunkt
        :type: daten.punkt.Punkt
        :param p_d: Dezimalstellen der eingegebenen Strecke
        :type: float """

        self.__p1: pkt.Punkt = copy.deepcopy(p_p1)
        self.__p2: pkt.Punkt = copy.deepcopy(p_p2)
        self.__d: float = p_d
        self.__nr: str=p_nr

    @classmethod
    def punkt_laenge(cls,p_p1: pkt.Punkt, p_laenge: float, p_nr=""):
        """alternativer Konstruktor
        :param p_p1: Punkt 1
        :type: daten.punkt.Punkt
        :param p_laenge: eingegebene Strecke
        :type: float
        :param p_nr: Punktnummer
        :type str"""

        #Dezimalstellen der Streckeneingabe holen
        p_d: float = len(str(p_laenge).split('.')[1])
        #Pseudopunkt auf x-Achse bilden
        p2:pkt.Punkt = pkt.Punkt(p_p1.hole_y(),p_p1.hole_x()+ p_laenge)
        return cls(p_p1, p2, p_d, p_nr)

    def setze_nr(self, p_nr: str):
        """setter in gon

        :param p_gon: Winkelgroesse in Gon
        :type: float
        """
        self.__nr = p_nr

    def hole_nr(self):
        """getter in gon
        :return: Winkelgroesse in Gon
        :rtype: float
        """
        return self.__nr

    def hole_p1(self) ->pkt.Punkt:
        """"Hole Startpunkt
        :return: Startpunkt
        :rtype: daten.punkt.Punkt
        """

        return self.__p1
    def hole_p2(self) ->pkt.Punkt:
        return self.__p2

    def riwi_laenge(self) -> Tuple[float, float]:
        """"Richtungswinkel und Strecke zurueckgeben
        :return: s12, t12
        :rtype: Tuple[float,float]
        
        """
        ts: Tuple[float,float]
        t,s=grundlagen.zweitegrund.umrechnen_koordinaten(self.__p1.hole_y(),self.__p1.hole_x(),self.__p2.hole_y(),self.__p2.hole_x())
        #Ausgabe des Richtungswinkels und der ggf. gerundeten Strecke
        return t,round(s, self.__d)
    def __str__(self) -> str:
        """Zeichenkette
        :return:Streckeninfos als Zeichenkette
        :rtype: str
        """
        ts: Tuple[float, float] = self.riwi_laenge()
        zeichen: str= "Nr:" + self.__nr + 'p1:' + str(self.__p1)+',p2:' + str(self.__p2) + ',t=' + str(ts[0]) + ',s=' + str(ts[1])
        return zeichen

if __name__=="__main__":
    p1: pkt.Punkt=pkt.Punkt(0.0,0.0, "p11")
    p2: pkt.Punkt=pkt.Punkt(1.0,1.0, "p12")
    s1: Strecke=Strecke.punkt_laenge(p1,5.0,"s1")
    s,t =s1.riwi_laenge()
    print(t)
    print(s)
    print(s1)
    pass
